transports/webrtc/src/connection/poll_data_channel.rs

```diff
@@ -0,0 +1,289 @@
+    }
+}
+
+impl AsyncWrite for PollDataChannel<'_> {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        let (fut, fut_is_new) = match self.write_fut.as_mut() {
+            Some(fut) => (fut, false),
+            None => {
+                let dc = self.data_channel.clone();
+                let bytes = Bytes::copy_from_slice(buf);
+                (
+                    self.write_fut
+                        .get_or_insert(Box::pin(async move { dc.write(&bytes).await })),
+                    true,
+                )
+            },
+        };
+
+        match fut.as_mut().poll(cx) {
+            Poll::Pending => {
+                // If it's the first time we're polling the future, `Poll::Pending` can't be
+                // returned because that would mean the `PollDataChannel` is not ready for writing. And
+                // this is not true since we've just created a future, which is going to write the
+                // buf to the underlying dc.
+                //
+                // It's okay to return `Poll::Ready` if the data is buffered (this is what the
+                // buffered writer and `File` do).
+                if fut_is_new {
+                    Poll::Ready(Ok(buf.len()))
+                } else {
+                    // If it's the subsequent poll, it's okay to return `Poll::Pending` as it
+                    // indicates that the `PollDataChannel` is not ready for writing. Only one future
+                    // can be in progress at the time.
+                    Poll::Pending
+                }
+            },
+            Poll::Ready(Err(e)) => {
+                self.write_fut = None;
+                Poll::Ready(Err(webrtc_error_to_io(e)))
+            },
+            Poll::Ready(Ok(n)) => {
+                self.write_fut = None;
+                Poll::Ready(Ok(n))
+            },
+        }
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        match self.write_fut.as_mut() {
+            Some(fut) => match fut.as_mut().poll(cx) {
+                Poll::Pending => Poll::Pending,
+                Poll::Ready(Err(e)) => {
+                    self.write_fut = None;
+                    Poll::Ready(Err(webrtc_error_to_io(e)))
+                },
+                Poll::Ready(Ok(_)) => {
+                    self.write_fut = None;
+                    Poll::Ready(Ok(()))
+                },
+            },
+            None => Poll::Ready(Ok(())),
+        }
+    }
+
+    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        let fut = match self.shutdown_fut.as_mut() {
+            Some(fut) => fut,
+            None => {
+                let dc = self.data_channel.clone();
+                self.shutdown_fut
+                    .get_or_insert(Box::pin(async move { dc.close().await }))
+            },
+        };
```

Comment:
If you use [`get_or_insert_with`](https://doc.rust-lang.org/std/option/enum.Option.html#method.get_or_insert_with), you should be able to get rid of the `match`:

```suggestion
	let dc = &self.data_channel;
        let fut = self.shutdown_fut.get_or_insert_with(|| {
        	let dc = dc.clone();
        	Box::pin(async move { dc.close().await })
        });
```